✨ Add k8s endpoints for inner cluster services

beranm14 opened the pull request:
Related: #36686

beranm14 committed:
✨ Add k8s endpoints for inner cluster services

Related: #36686

tomashejatko commented on elasticsearch.yml.tpl:
```diff
@@ -21,4 +21,6 @@ discovery:
   zen:
     hosts_provider: gce
 
+cluster.initial_master_nodes: [${nodes}]
```
Jak se tohle zachová po restartu celého clusteru? Jakože všechny node down a všechny up? Nečetl jsem k tomu docku, takže popravdě furt moc nechápu jak tento parametr funguje, ale zdá se mi divné že by tam člověk jenom vylistoval nody a tím bylo hotovo.

beranm14 replied:
Bohužel bez toho to nenahodíš :-| háže to:
```
[2020-04-07T15:19:51,343][WARN ][o.e.c.c.ClusterFormationFailureHelper] [elasticsearch-prod-0] master not discovered yet, this node has not previously joined a bootstrapped (v7+) cluster, and [cluster.initial_master_nodes] is empty on this node: have discovered [{elasticsearch-prod-0}{nlgXLwvcSnaLPLDc4B4nhA}{VA4M1ZAxTwCuKnwiU5Pp2w}{10.156.0.58}{10.156.0.58:9300}{dilm}{ml.machine_memory=3869970432, xpack.installed=true, ml.max_open_jobs=20}, {elasticsearch-prod-2}{XIYmNokfRe-vaFMXdio92A}{2tDSSoQ-Sk-txf6ebKoAQQ}{10.156.0.57}{10.156.0.57:9300}{dilm}{ml.machine_memory=3869970432, ml.max_open_jobs=20, xpack.installed=true}, {elasticsearch-prod-1}{BXFtnvvPRsmI4QSpvXI_lw}{mCCpj-yuQQelspKXzHQsEw}{10.156.0.59}{10.156.0.59:9300}{dilm}{ml.machine_memory=3869986816, ml.max_open_jobs=20, xpack.installed=true}]; discovery will continue using [127.0.0.1:9301, 127.0.0.1:9302, 127.0.0.1:9303, 127.0.0.1:9304, 127.0.0.1:9305, [::1]:9301, [::1]:9302, [::1]:9303, [::1]:9304, [::1]:9305, 10.156.0.59:9300, 10.156.0.57:9300, 10.156.15.233:9300] from hosts providers and [{elasticsearch-prod-0}{nlgXLwvcSnaLPLDc4B4nhA}{VA4M1ZAxTwCuKnwiU5Pp2w}{10.156.0.58}{10.156.0.58:9300}{dilm}{ml.machine_memory=3869970432, xpack.installed=true, ml.max_open_jobs=20}] from last-known cluster state; node term 0, last-accepted version 0 in term 0
```
Na to nastavení jsem přišel googlením a taky moc nevím, jak se to chová. Můžeme to samozřejmě otestovat, jestli chceš.

tomashejatko commented on example/terraform_backend.sh:
```diff
@@ -0,0 +1,11 @@
+export TF_VAR_region=europe-west3
+export TF_VAR_zone=europe-west3-c
+export TF_VAR_vault_secret_path=secret/devops/internal/hejda-test
+export VAULT_ADDR=https://vault01.ack.ee
```
Tohle není rozumné úplně veřejně vystavovat, zvlášť když jsme asi o rok pozadu s verzí Vaultu

beranm14 replied:
To sem nepiš :-D to nikdo nemusí vědět

tomashejatko commented on providers.tf:
```diff
@@ -0,0 +1,19 @@
+provider "kubernetes" {
+  version = "~> 1.10.0"
```
Můžeme upgradnout na `~> 1.11.0` ?

tomashejatko committed:
Use GCE as seed provider, should fix bootstrapping problems.

Changes according to https://www.elastic.co/guide/en/elasticsearch/plugins/current/discovery-gce-usage.html

tomashejatko committed:
⚡️ Fix GCE discovery plugins settings, as it works as seed provider, we can also remove initial_master_nodes setting

Related: #36686

beranm14 committed:
✨ Add k8s as optional part, introduced gcp internal load balancer for es service

Related: #36686

beranm14 committed:
✨ Add bootstrap.sh changes to workout issues with SE bootstrap master election

Related: #36686

sandratatarevicova commented on bootstrap.sh:
```diff
@@ -1,10 +1,77 @@
+
+fi
+
+su elasticsearch -s /bin/bash -c '/usr/share/elasticsearch/bin/elasticsearch | tee /tmp/elasticsearch_bootstrap_start.log >/dev/full' &
```
Proč se tu přesměrovává výstup do `/dev/full`?

sandratatarevicova commented on bootstrap.sh:
```diff
@@ -1,10 +1,77 @@
+
+echo "waiting for ES7 bootstrap to start"
+while true; do
+  grep "starting GCE discovery service" /tmp/elasticsearch_bootstrap_start.log && break;
```
Tohle bych přepsala na něco jako:

```bash
while grep -v "starting GCE discovery service" /tmp/elasticsearch_bootstrap_start.log; do
  sleep 10
done
```

A asi bych se podívala, jestli na ten start neumíme počkat lepším způsobem než grepováním logu.

sandratatarevicova replied:
Nedalo by se použít třeba https://www.elastic.co/guide/en/elasticsearch/reference/7.7/cat-health.html?

beranm14 replied:
V týhle části ještě ne, protože se to odmítne zapnout a api ještě nic nevrací. To když už, tak na řádku 85

sandratatarevicova commented on bootstrap.sh:
```diff
@@ -1,10 +1,77 @@
+  sleep 10
+done
+
+sleep 10
```
Nevím, proč je tu potřeba tenhle `sleep`, ale na první pohled mi to přijde jako další důvod, proč nečekat na základě grepování logu 🙂

sandratatarevicova commented on bootstrap.sh:
```diff
@@ -1,10 +1,77 @@
+  sleep 10
+done
+
+while pgrep java; do
```
Proč tu zabíjíme javu?

sandratatarevicova replied:
> no matter what happened before, elasticsearch needs to be stopped to be start later on from systemd service

Nemůžeme nějak standardně zabít přímo Elastic? Vždyt tam má svou systemd unit, ne?

sandratatarevicova commented on bootstrap.sh:
```diff
@@ -1,10 +1,77 @@
+fi
+if [[ -f /tmp/elasticsearch_bootstrap_master_list_start.log ]]; then
+  rm /tmp/elasticsearch_bootstrap_master_list_start.log
+fi
```
Bez toho, abych to celé důkladně zkoumala těm změnám v `bootstrap.sh` moc nerozumím.

beranm14 replied:
Tohle není zásadní, to je jen poklizení souborů, kam se ukládali logy pro grep

sandratatarevicova replied:
To byl komentář k celému bootstrap.sh 🙂
